### PR TITLE
Support generating android adaptive icons

### PR DESCRIPTION
**Pull request by vigorox**
Support generating android adaptive icons:
```
flavorizr:
  flavors:
    flavor_1:
      app:
        name: "myApp"
      android:
        applicationId: "com.xxx.myApp"
        icon: "assets/icon.png"
        iconForeground: "assets/foreground.png"
        iconBackground: "assets/background.png"
```

The instruction to generate adaptive icons is "**android:adaptiveIcons**"

### CODE REVIEW

**Comment by AngeloAvv** [comment]
Hey @vigorox , thanks for supporting this project. I have a few questions for you:
* What if we avoid creating a redundant processor for adaptive icons and we put everything into the default `android:icons` step instead?
* I do agree with having a separate configuration for adaptive icons, what I personally don't like is letting the Android model grow indefinitely: I would create a key named `adaptiveIcon` which replaces both iconForeground and iconBackground, and then we can move the foreground and background fields into it as follows

``` groovy
flavorizr:
  flavors:
    flavor_1:
      app:
        name: "myApp"
      android:
        applicationId: "com.xxx.myApp"
        icon: "assets/icon.png"
        adaptiveIcon:
          foreground: "assets/foreground.png"
          background: "assets/background.png"
```

* I'm confused about having path processors, what's the purpose of them? I think you can easily customize the behavior of NewFileStringProcessor to create nested folders if they don't exist. Please triple-check everything to avoid breaking existing processors :sweat_smile: 
* I would avoid deleting previously existing icons as not everyone may expect this behavior: adding more lines in the README so people are aware that they should delete these icons manually instead. 

What do you think? Thank you!

**Comment by vigorox** [comment]
Hi AngeloAvv,

I modified based on your suggestions.
Please take a look.

Thanks,

File: README.md
Comment (AngeloAvv) [inline README.md:1]: I think you can rollback this file: actually is not perfectly formatted

File: lib/src/parser/models/flavors/android.dart
Comment (AngeloAvv) [inline lib/src/parser/models/flavors/android.dart:49]:
```suggestion
  final AdaptiveIcon? adaptiveIcon;
```

Sorry I haven't been clear enough: If we create a nested model we have more control over the fields.

File: lib/src/parser/models/flavors/android.dart
Comment (AngeloAvv) [inline lib/src/parser/models/flavors/android.dart:59]: Then this could be nullable

File: lib/src/processors/processor.dart
Comment (AngeloAvv) [inline lib/src/processors/processor.dart:201]: Instead of having a QueueProcessor here, you can collapse the `AndroidAdaptiveIconProcessor` behaviour into the `AndroidIconsProcessor` using the spread operator over the super constructor:

``` dart
...config.androidFlavors
     .where((_, flavor) =>
         flavor.android!.adaptiveIcon!['foreground'] != null &&
         flavor.android!.adaptiveIcon!['background'] != null)
     .map((flavorName, flavor) => MapEntry(
           flavorName,
           QueueProcessor([
             AndroidAdaptiveIconXmlProcessor(flavorName,
                 config: config),
             AndroidAdaptiveIconProcessor(
               flavor.android!.adaptiveIcon!['foreground'] ?? '',
               flavor.android!.adaptiveIcon!['background'] ?? '',
               flavorName,
               config: config,
             ),
           ], config: config),
         ))
     .values,
```

Also, I think it's better to extend classes starting from QueueProcessors instead of using them in this way IMHO.

File: lib/src/processors/android/icons/android_adaptive_icons_processor.dart
Comment (AngeloAvv) [inline lib/src/processors/android/icons/android_adaptive_icons_processor.dart:1]: This could be deleted after merging the behaviour into android_icons_processor.dart

File: lib/src/processors/commons/abstract_file_string_processor.dart
Comment (AngeloAvv) [inline lib/src/processors/commons/abstract_file_string_processor.dart:47]: This is the solution I was mentioning before:
https://github.com/AngeloAvv/flutter_flavorizr/pull/219/files

File: pubspec.yaml
Comment (AngeloAvv) [inline pubspec.yaml:24]: I think this is no longer needed

**Review by AngeloAvv — CHANGES_REQUESTED** [review]
Nice! I added a few comments to enhance code readability

**Comment by AngeloAvv** [comment]
@vigorox please don't squash/force-push your commits, I'll squash them once everything is ready

File: lib/src/parser/models/flavors/android/adaptive_icon.dart
Comment (AngeloAvv) [inline lib/src/parser/models/flavors/android/adaptive_icon.dart:8]: Those should be mandatory when an adaptive icon is defined, aren't they?

File: lib/src/processors/android/icons/android_icons_processor.dart
Comment (AngeloAvv) [inline lib/src/processors/android/icons/android_icons_processor.dart:71]: You can get rid of the null-coalescing operator when you change the `background` and `foreground` attributes to non-null.

**Review by AngeloAvv — CHANGES_REQUESTED** [review]
Okay! Everything looks good to me :rocket: 

Two final tweaks and we're ready to go!

**Comment by vigorox** [comment]
Hi AngeloAvv,

I refactored the code based on your review.

If any further change is required, please tell me.

Thanks,

**Review by AngeloAvv — CHANGES_REQUESTED** [review]
Hey @vigorox , sorry for my late reply. I haven't had enough time to take a look at your latest changes until now.

I found a final thing that needs to be addressed (sorry for being so picky :laughing: ) and I think we're good to go

File: lib/src/processors/android/icons/android_adaptive_icon_processor.dart
Comment (vigorox) [reply lib/src/processors/android/icons/android_adaptive_icon_processor.dart]: @AngeloAvv,

The code of processors is modified based on your review.

**Comment by kbudulski** [comment]
@AngeloAvv That's a cool feature and I see few PRs referencing it, where can we expect a new release with adaptive icons included? 🙂

**Comment by AngeloAvv** [comment]
Hey @kbudulski , I was waiting for feedback from [this](https://github.com/AngeloAvv/flutter_flavorizr/issues/212) before releasing a new version of flutter_flavorizr

**Pull request merged**
